File: bindings/vbnet/src/EveraniumItb.VisualBasic/Pipeline.vb
' Delegating lifetime wrapper around the C# layer's Pipeline.
'
' Every member forwards to the sibling C# binding (bindings/csharp)
' through plain CLR bytecode interop — no FFI layer of its own. The
' C# side carries the P/Invoke surface, the handle lifetime
' (SafeHandle finalizer backstop), the buffer pre-allocation with
' the BufferTooSmall retry-once, and the libitb lookup order; this
' layer re-shapes that surface into idiomatic VB.NET (Using blocks,
' Try ... Catch ex As ItbException, Byte() signatures).

''' <summary>
''' A Triple Pipeline session.
'''
''' <see cref="Save"/> exports the self-describing session blob the
''' receiver feeds to <see cref="Load"/> / <see cref="LoadF"/>;
''' <see cref="Rekey"/> refreshes it. Disposing the Pipeline frees
''' the handle (libitb zeroes key material internally); an
''' undisposed Pipeline is reclaimed by the C# layer's SafeHandle
''' finalizer. Profile records (<see cref="Inspect"/> /
''' <see cref="Lookup"/> results, <see cref="Register"/> input) are
''' the C# layer's <see cref="Global.Itb.Profile"/> type.
'''
''' Streaming-decrypt caveat: chunked Streaming AEAD verifies per
''' chunk, so plaintext of verified chunks is released before a later
''' chunk can fail authentication.
''' </summary>
Public NotInheritable Class Pipeline
    Implements IDisposable

    Private ReadOnly _inner As Global.Itb.Pipeline

    Private Sub New(inner As Global.Itb.Pipeline)
        _inner = inner
    End Sub

    ''' <summary>The wrapped C# Pipeline the stream sessions run
    ''' against.</summary>
    Friend ReadOnly Property Inner As Global.Itb.Pipeline
        Get
            Return _inner
        End Get
    End Property

    ''' <summary>Constructs a fresh Pipeline against the named
    ''' profile. The session blob is available through
    ''' <see cref="Save"/>.</summary>
    Public Shared Function Init(profile As String, Optional opts As Opts = Nothing) As Pipeline
        Return Guarded(Function() New Pipeline(Global.Itb.Pipeline.Init(profile, InnerOpts(opts))))
    End Function

    ''' <summary>
    ''' Reconstructs a Pipeline from a blob produced by
    ''' <see cref="Save"/> or <see cref="Rekey"/>. The blob's embedded
    ''' profile record is the sole structural source. Omitting
    ''' <paramref name="permMaster"/> / <paramref name="wrapMaster"/>
    ''' uses the blob-embedded masters; supplying both (non-empty)
    ''' overrides them. Supplying only one raises
    ''' <see cref="ArgumentException"/>.
    ''' </summary>
    Public Shared Function Load(
            blob As Byte(),
            Optional permMaster As Byte() = Nothing,
            Optional wrapMaster As Byte() = Nothing) As Pipeline
        Return Guarded(Function() New Pipeline(Global.Itb.Pipeline.Load(
            blob, permMaster, wrapMaster)))
    End Function

    ''' <summary><see cref="Load"/> for a blob stored in a file; the
    ''' file is read inside the library. Same masters
    ''' semantics.</summary>
    Public Shared Function LoadF(
            path As String,
            Optional permMaster As Byte() = Nothing,
            Optional wrapMaster As Byte() = Nothing) As Pipeline
        Return Guarded(Function() New Pipeline(Global.Itb.Pipeline.LoadF(
            path, permMaster, wrapMaster)))
    End Function

    ''' <summary>Decodes the blob's embedded profile record without
    ''' opening a Pipeline. No registry read, no primitive
    ''' probe.</summary>
    Public Shared Function Inspect(blob As Byte()) As Global.Itb.Profile
        Return Guarded(Function() Global.Itb.Pipeline.Inspect(blob))
    End Function

    ''' <summary>
    ''' Registers <paramref name="profile"/> under
    ''' <paramref name="name"/> so subsequent <see cref="Init"/> /
    ''' <see cref="Lookup"/> calls resolve it. Every field rule is
    ''' validated by Go; a duplicate name fails with
    ''' <see cref="Status.ProfileExists"/>.
    ''' </summary>
    Public Shared Sub Register(name As String, profile As Global.Itb.Profile)
        Guarded(Sub() Global.Itb.Pipeline.Register(name, profile))
    End Sub

    ''' <summary>Looks up a registered profile (shipped or
    ''' <see cref="Register"/>ed) by name; an unknown name fails with
    ''' <see cref="Status.UnknownProfile"/>.</summary>
    Public Shared Function Lookup(name As String) As Global.Itb.Profile
        Return Guarded(Function() Global.Itb.Pipeline.Lookup(name))
    End Function

    ''' <summary>The sorted names of every registered
    ''' profile.</summary>
    Public Shared Function Profiles() As String()
        Return Guarded(Function() Global.Itb.Pipeline.Profiles())
    End Function

    ''' <summary>The current self-describing session blob: the bytes
    ''' <see cref="Init"/> produced, the bytes <see cref="Load"/>
    ''' re-marshalled, or the bytes of the latest
    ''' <see cref="Rekey"/>.</summary>
    Public Function Save() As Byte()
        Return Guarded(Function() _inner.Save())
    End Function

    ''' <summary>Writes <see cref="Save"/> to <paramref name="path"/>
    ''' inside the library with mode 0600; the containing directory
    ''' must exist.</summary>
    Public Sub SaveF(path As String)
        Guarded(Sub() _inner.SaveF(path))
    End Sub

    ''' <summary>Sets the worker cap for every subsequent cipher call.
    ''' <paramref name="n"/> is clamped, never rejected: n &lt;= 0
    ''' selects auto (CPU count), n &gt; 256 is treated as 256. Only
    ''' the handle statuses raise.</summary>
    Public Sub MaxWorkers(n As Integer)
        Guarded(Sub() _inner.MaxWorkers(n))
    End Sub

    ''' <summary>
    ''' Rotates the parallax + wrapper masters and returns the fresh
    ''' session blob (also available through <see cref="Save"/>).
    ''' Must not run concurrently with cipher calls or open stream
    ''' sessions on the same Pipeline.
    ''' </summary>
    Public Function Rekey(permMaster As Byte(), wrapMaster As Byte()) As Byte()
        Return Guarded(Function() _inner.Rekey(permMaster, wrapMaster))
    End Function

    ''' <summary>
    ''' Zeroes the Pipeline's key material and marks it closed.
    ''' Idempotent; subsequent cipher calls fail with
    ''' <see cref="Status.TripleClosed"/>. The handle itself is
    ''' released by <see cref="Dispose"/>.
    ''' </summary>
    Public Sub Close()
        Guarded(Sub() _inner.Close())
    End Sub

    ''' <summary>Single Message encrypt: one call, one self-contained
    ''' wire.</summary>
    Public Function EncryptMessage(plaintext As Byte()) As Byte()
        Return Guarded(Function() _inner.EncryptMessage(plaintext))
    End Function

    ''' <summary>Receive-side counterpart of
    ''' <see cref="EncryptMessage"/>.</summary>
    Public Function DecryptMessage(wire As Byte()) As Byte()
        Return Guarded(Function() _inner.DecryptMessage(wire))
    End Function

    ''' <summary>
    ''' One-shot stream encrypt for callers holding the whole
    ''' plaintext in memory. For bounded-memory streaming use
    ''' <see cref="BeginEncryptStream"/> /
    ''' <see cref="EncryptStreamPump"/>.
    ''' </summary>
    Public Function EncryptStreamOneShot(plaintext As Byte()) As Byte()
        Return Guarded(Function() _inner.EncryptStreamOneShot(plaintext))
    End Function

    ''' <summary>Receive-side counterpart of
    ''' <see cref="EncryptStreamOneShot"/>.</summary>
    Public Function DecryptStreamOneShot(wire As Byte()) As Byte()
        Return Guarded(Function() _inner.DecryptStreamOneShot(wire))
    End Function

    ''' <summary>Opens an incremental encrypt session (plaintext in,
    ''' wire out).</summary>
    Public Function BeginEncryptStream() As EncryptStream
        Return New EncryptStream(Me, Guarded(Function() _inner.BeginEncryptStream()))
    End Function

    ''' <summary>Opens an incremental decrypt session (wire in,
    ''' plaintext out).</summary>
    Public Function BeginDecryptStream() As DecryptStream
        Return New DecryptStream(Me, Guarded(Function() _inner.BeginDecryptStream()))
    End Function

    ''' <summary>
    ''' Pumps <paramref name="source"/> through an encrypt session
    ''' into <paramref name="destination"/> with bounded memory: feed
    ''' a block, drain available wire, repeat; end + final drain on
    ''' source EOF. The session is freed on return.
    ''' </summary>
    Public Sub EncryptStreamPump(source As IO.Stream, destination As IO.Stream)
        Guarded(Sub() _inner.EncryptStreamPump(source, destination))
    End Sub

    ''' <summary>Receive-side counterpart of
    ''' <see cref="EncryptStreamPump"/>.</summary>
    Public Sub DecryptStreamPump(source As IO.Stream, destination As IO.Stream)
        Guarded(Sub() _inner.DecryptStreamPump(source, destination))
    End Sub

    Public Sub Dispose() Implements IDisposable.Dispose
        _inner.Dispose()
    End Sub

    Private Shared Function InnerOpts(opts As Opts) As Global.Itb.Opts
        Return If(opts Is Nothing, Nothing, opts.Inner)
    End Function
End Class

File: bindings/scala/src/main/scala/dev/everanium/itb/Pipeline.scala
// Idiomatic Scala lifetime + error surface over the Java binding's
// Triple Pipeline.

package dev.everanium.itb

import java.io.{InputStream, OutputStream}

import scala.jdk.CollectionConverters.*

import com.everanium.itb.Pipeline as JPipeline

import ItbError.attempt

/** Result of one stream drain call: `count` bytes were placed into
  * the destination buffer; `finished` marks the session output as
  * complete.
  */
final case class ReadResult(count: Int, finished: Boolean)

/** A Triple Pipeline session.
  *
  * [[save]] exports the self-describing session blob the receiver
  * feeds to [[Pipeline.load]] / [[Pipeline.loadF]]; [[rekey]]
  * refreshes it. `close()` releases the native handle (libitb zeroes
  * key material internally), so the class composes with
  * `scala.util.Using.resource`; the Java binding's finalization
  * backstop reclaims an unclosed Pipeline.
  *
  * Every fallible operation returns `Either[ItbError, _]`; the
  * for-comprehension is the intended composition style. Callers
  * preferring exceptions can `fold(throw _, identity)` — the error
  * value is a `RuntimeException`.
  *
  * Streaming-decrypt caveat: chunked Streaming AEAD verifies per
  * chunk, so plaintext of verified chunks is released before a later
  * chunk can fail authentication.
  */
final class Pipeline private (private[itb] val underlying: JPipeline)
    extends AutoCloseable:

  /** The current self-describing session blob: the bytes [[Pipeline.init]]
    * produced, the bytes [[Pipeline.load]] re-marshalled, or the
    * bytes of the latest [[rekey]].
    */
  def save(): Either[ItbError, Array[Byte]] = attempt(underlying.save())

  /** Writes [[save]] to `path` inside the library with mode 0600;
    * the containing directory must exist.
    */
  def saveF(path: String): Either[ItbError, Unit] = attempt(underlying.saveF(path))

  /** Sets the worker cap for every subsequent cipher call. `n` is
    * clamped, never rejected: `n <= 0` selects auto (CPU count),
    * `n > 256` is treated as 256. Only the handle statuses fail.
    */
  def maxWorkers(n: Int): Either[ItbError, Unit] = attempt(underlying.maxWorkers(n))

  /** Rotates the parallax + wrapper masters and returns the fresh
    * session blob (also available through [[save]]). Must not run
    * concurrently with cipher calls or open stream sessions on the
    * same Pipeline.
    */
  def rekey(permMaster: Array[Byte], wrapMaster: Array[Byte]): Either[ItbError, Array[Byte]] =
    attempt(underlying.rekey(permMaster, wrapMaster))

  /** Zeroes the Pipeline's key material and marks it closed.
    * Idempotent; subsequent cipher calls fail with
    * [[Status.TripleClosed]]. The native handle itself is released
    * by [[close]].
    */
  def closeSession(): Either[ItbError, Unit] = attempt(underlying.destroy())

  /** Single Message encrypt: one call, one self-contained wire. */
  def encryptMessage(plaintext: Array[Byte]): Either[ItbError, Array[Byte]] =
    attempt(underlying.encryptMessage(plaintext))

  /** Receive-side counterpart of [[encryptMessage]]. */
  def decryptMessage(wire: Array[Byte]): Either[ItbError, Array[Byte]] =
    attempt(underlying.decryptMessage(wire))

  /** One-shot stream encrypt for callers holding the whole plaintext
    * in memory. For bounded-memory streaming use
    * [[beginEncryptStream]] / [[encryptStreamPump]].
    */
  def encryptStreamOneShot(plaintext: Array[Byte]): Either[ItbError, Array[Byte]] =
    attempt(underlying.encryptStreamOneShot(plaintext))

  /** Receive-side counterpart of [[encryptStreamOneShot]]. */
  def decryptStreamOneShot(wire: Array[Byte]): Either[ItbError, Array[Byte]] =
    attempt(underlying.decryptStreamOneShot(wire))

  /** Opens an incremental encrypt session (plaintext in, wire out).
    * The session holds a reference to this Pipeline, keeping it
    * reachable while the session is live.
    */
  def beginEncryptStream(): Either[ItbError, EncryptStream] =
    attempt(new EncryptStream(this, underlying.encryptStream()))

  /** Opens an incremental decrypt session (wire in, plaintext out).
    * The session holds a reference to this Pipeline, keeping it
    * reachable while the session is live.
    */
  def beginDecryptStream(): Either[ItbError, DecryptStream] =
    attempt(new DecryptStream(this, underlying.decryptStream()))

  /** Pumps `source` through an encrypt session into `destination`
    * with bounded memory: feed a block, drain available wire,
    * repeat; end + final drain on source EOF. The session is freed
    * on return.
    */
  def encryptStreamPump(source: InputStream, destination: OutputStream): Either[ItbError, Unit] =
    attempt(underlying.encryptStreamPump(source, destination))

  /** Receive-side counterpart of [[encryptStreamPump]]. */
  def decryptStreamPump(source: InputStream, destination: OutputStream): Either[ItbError, Unit] =
    attempt(underlying.decryptStreamPump(source, destination))

  /** Releases the native handle. Idempotent. */
  override def close(): Unit = underlying.close()

object Pipeline:

  /** Constructs a fresh Pipeline against the named profile. */
  def init(profile: String, opts: Opts = Opts.empty): Either[ItbError, Pipeline] =
    attempt(new Pipeline(JPipeline.init(profile, opts.toJava)))

  /** Reconstructs a Pipeline from a blob produced by
    * [[Pipeline.save]] or [[Pipeline.rekey]], using the
    * blob-embedded masters. The blob's embedded profile record is the
    * sole structural source. See [[loadWithMasters]] to override the
    * masters.
    */
  def load(blob: Array[Byte]): Either[ItbError, Pipeline] =
    attempt(new Pipeline(JPipeline.load(blob)))

  /** [[load]] with explicit (non-empty) parallax + wrapper masters
    * overriding the blob-embedded ones; both must be supplied.
    */
  def loadWithMasters(
      blob: Array[Byte],
      permMaster: Array[Byte],
      wrapMaster: Array[Byte]
  ): Either[ItbError, Pipeline] =
    attempt(new Pipeline(JPipeline.load(blob, permMaster, wrapMaster)))

  /** [[load]] for a blob stored in a file; the file is read inside
    * the library.
    */
  def loadF(path: String): Either[ItbError, Pipeline] =
    attempt(new Pipeline(JPipeline.loadF(path)))

  /** [[loadF]] with explicit (non-empty) parallax + wrapper masters
    * overriding the blob-embedded ones; both must be supplied.
    */
  def loadFWithMasters(
      path: String,
      permMaster: Array[Byte],
      wrapMaster: Array[Byte]
  ): Either[ItbError, Pipeline] =
    attempt(new Pipeline(JPipeline.loadF(path, permMaster, wrapMaster)))

  /** Decodes the blob's embedded profile record without opening a
    * Pipeline. No registry read, no primitive probe.
    */
  def inspect(blob: Array[Byte]): Either[ItbError, Profile] =
    attempt(JPipeline.inspect(blob))

  /** Registers `profile` under `name` so subsequent [[init]] /
    * [[lookup]] calls resolve it. Every field rule is validated by
    * Go; a duplicate name fails with [[Status.ProfileExists]].
    */
  def register(name: String, profile: Profile): Either[ItbError, Unit] =
    attempt(JPipeline.register(name, profile))

  /** Looks up a registered profile (shipped or [[register]]ed) by
    * name; an unknown name fails with [[Status.UnknownProfile]].
    */
  def lookup(name: String): Either[ItbError, Profile] =
    attempt(JPipeline.lookup(name))

  /** The sorted names of every registered profile. */
  def profiles(): Either[ItbError, List[String]] =
    attempt(JPipeline.profiles().asScala.toList)
